src/main/java/net/pms/database/TableFilesStatus.java

```diff
@@ -181,11 +181,12 @@ public static void setDirectoryFullyPlayed(final String fullPathToFile, final bo
 			TABLE_LOCK.writeLock().lock();
 			try (Statement statement = connection.createStatement()) {
 				try (ResultSet result = statement.executeQuery(query)) {
-					TABLE_LOCK.writeLock().unlock();
 					while (result.next()) {
```

Comment:
This is wrong, you should never release the lock while you're iterating a result. That means that other threads could alter the table in the meanwhile rendering the ResultSet you're iterating invalid. You want an operation like this to be "atomic" in the sense that the operation is done without concurrent changes by other parties, and thus the lock should be held from you start the operation until it's finished or fails. The lock is a ```ReentrantReadWriteLock``` meaning that it's ok for the same thread to aquire the lock recursively. As long as it already holds the lock, it will always be granted the recursive lock. I'm guessing this is the reason you thought you had to release it before aquiring it again in ```setFullyPlayed```. 

My "best practices" tells me that it's ok to aquire a read lock while holding a write lock, but not the other way. I never ask for a write lock while holding a read lock. 

When I think about it now, I'm not sure if that's actually necessary here - I'd think that aquiring a write lock while holding a read lock would just wait until all other read- and write locks were released and then proceed, but I'd have to make really sure that's the way it works before changing my "best practice". All I know is that I've been burned by doing that before, but that was in a different language with a different locking mechanism so it might not apply here. I guess I'll have to investigate that when the need to do that present itself, in the meantime I'll just avoid doing it.